flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala

```diff
@@ -0,0 +1,141 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.runtime.aggregate
+
+import org.apache.flink.api.common.state.{ListState, ListStateDescriptor}
+import org.apache.flink.api.java.typeutils.RowTypeInfo
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.runtime.state.{FunctionInitializationContext, FunctionSnapshotContext}
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction
+import org.apache.flink.streaming.api.functions.ProcessFunction
+import org.apache.flink.table.functions.{Accumulator, AggregateFunction}
+import org.apache.flink.types.Row
+import org.apache.flink.util.{Collector, Preconditions}
+import org.apache.flink.api.common.state.ValueState
+import org.apache.flink.api.common.state.ValueStateDescriptor
+import scala.util.control.Breaks._
+
+/**
+  * Process Function used for the aggregate in partitioned bounded windows in
+  * [[org.apache.flink.streaming.api.datastream.DataStream]]
+  *
+  * @param aggregates the list of all [[org.apache.flink.table.functions.AggregateFunction]]
+  *                   used for this aggregation
+  * @param aggFields  the position (in the input Row) of the input value for each aggregate
+  * @param forwardedFieldCount Is used to indicate fields in the current element to forward
+  * @param rowTypeInfo Is used to indicate the field schema
+  * @param time_boundary Is used to indicate the processing time boundaries
+  */
+class ProcTimeBoundedProcessingOverProcessFunction(
+    private val aggregates: Array[AggregateFunction[_]],
+    private val aggFields: Array[Int],
+    private val forwardedFieldCount: Int,
+    private val rowTypeInfo: RowTypeInfo,
+    private val time_boundary: Long)
+  extends ProcessFunction[Row, Row] {
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+
+  private var accumulators: Row = _
+  private var output: Row = _
+  private var windowBuffer: ListState[Tuple2[Long,Row]] = null
+  private var state: ValueState[Row] = _
+
+  
+  override def open(config: Configuration) {
+    output = new Row(forwardedFieldCount + aggregates.length)
+    
+    accumulators = new Row(aggregates.length)
+    var i = 0
+    while (i < aggregates.length) {
+        accumulators.setField(i, aggregates(i).createAccumulator())
+        i += 1
+      } 
+    
+    // We keep the elements received in a list state 
+    // together with the ingestion time in the operator
+    val bufferDescriptor: ListStateDescriptor[Tuple2[Long,Row]] = 
+    new ListStateDescriptor[Tuple2[Long,Row]]("windowBufferState", classOf[Tuple2[Long,Row]])
+    windowBuffer = getRuntimeContext.getListState(bufferDescriptor)
+
+    val stateDescriptor: ValueStateDescriptor[Row] =
+    new ValueStateDescriptor[Row]("overState", classOf[Row] , accumulators)      
```

Comment:
Updating the value directly is not possible, because we would need to initialize the state for all keys. 

Also, a ValueState with default value is also first retrieving the value from the state, checks if its `null` and returns either the non-null value or checks if a default value has been configured or not and returns either that or `null`. So it has the same overhead as "manually" checking in the `processElement()`.